grug/types/src/length_bounded.rs

```diff
@@ -0,0 +1,371 @@
+use {
+    crate::{Binary, StdError, StdResult},
+    borsh::{BorshDeserialize, BorshSerialize},
+    grug_math::Inner,
+    serde::{Deserialize, Deserializer, Serialize, Serializer},
+    std::{
+        collections::{BTreeMap, BTreeSet},
+        ops::Deref,
+    },
+};
+
+pub type MaxLength<T, const MAX: usize> = LengthBounded<T, 0, MAX>;
+pub type MinLength<T, const MIN: usize> = LengthBounded<T, MIN, { usize::MAX }>;
+pub type NonEmpty<T> = MinLength<T, 1>;
+pub type FixedLength<T, const LEN: usize> = LengthBounded<T, LEN, LEN>;
+pub type ByteArray<const LEN: usize> = FixedLength<Binary, LEN>;
+
+/// A wrapper that enforces the value to be within the specified length.
+/// The value must implement the `LengthBounds` trait.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct LengthBounded<T, const MIN: usize, const MAX: usize>(T)
+where
+    T: LengthBounds;
+
+pub trait LengthBounds {
```

Comment:
Can you rename this to `Lengthy`?

`LengthBounds` doesn't make sense because it isn't bounds